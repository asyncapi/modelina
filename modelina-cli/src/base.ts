import { Command } from '@oclif/core';

export default abstract class extends Command {
  async catch(err: Error & { exitCode?: number; }): Promise<any> {
    try {
      return await super.catch(err);
    } catch (e: any) {
      if (e instanceof Error) {
        this.logToStderr(`${e.name}: ${e.message}`);
        process.exitCode = 1;
      }
    }
  }
}
