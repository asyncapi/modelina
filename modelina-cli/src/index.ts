export {run} from '@oclif/core'

export {generateModels, ModelinaArgs, ModelinaFlags, Languages} from './helpers/generate'