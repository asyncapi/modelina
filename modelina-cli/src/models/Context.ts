import { readFileSync, promises as fs, existsSync, lstatSync } from 'fs';
import * as path from 'path';
import * as os from 'os';

import {
  ContextNotFoundError,
  MissingContextFileError,
  MissingCurrentContextError,
  ContextFileWrongFormatError,
  ContextAlreadyExistsError,
  ContextFileEmptyError,
  ContextFileWriteError,
} from '../errors/context-error';

const { readFile, writeFile } = fs;

// `repoRootPath` is optimistically assigned current working directory's
// filesystem path because chances are it will become 'official' repository root
// down the execution.
//
// `REPO_ROOT_PATH` will be converted to a real constant after migration of the
// codebase to ES2022 or higher and introduction of construction
//
// const REPO_ROOT_PATH = await getRepoRootPath(process.cwd());
//
// See explanation of the situation with `CONTEXT_FILE_PATH` below.
let REPO_ROOT_PATH = process.cwd();
getRepoRootPath(process.cwd());

const DEFAULT_CONTEXT_FILENAME = '.asyncapi-cli';
const DEFAULT_CONTEXT_FILE_LOCATION = os.homedir();
export const DEFAULT_CONTEXT_FILE_PATH = path.resolve(
  DEFAULT_CONTEXT_FILE_LOCATION,
  DEFAULT_CONTEXT_FILENAME
);

const CONTEXT_FILENAME =
  process.env.CUSTOM_CONTEXT_FILENAME || DEFAULT_CONTEXT_FILENAME;
const CONTEXT_FILE_LOCATION =
  process.env.CUSTOM_CONTEXT_FILE_LOCATION || DEFAULT_CONTEXT_FILE_LOCATION;

// Usage of promises for assignment of their resolved values to constants is
// known to be troublesome:
// https://www.reddit.com/r/learnjavascript/comments/p7p7zw/assigning_data_from_a_promise_to_a_constant
//
// In this particular case and usage of ES6, there is a race condition during
// code execution, due to faster assignment of default values to
// `CONTEXT_FILE_PATH` than resolution of the promise. This is the cause
// `CONTEXT_FILE_PATH` will always pick default values for context file's path
// instead of waiting for resolution of the promise from `getContextFilePath()`.
// The situation might become better with use of top-level await which should
// pause code execution, until promise in construction
//
// const CONTEXT_FILE_PATH = await getContextFilePath() || path.resolve(CONTEXT_FILE_LOCATION, CONTEXT_FILENAME) || DEFAULT_CONTEXT_FILE_PATH;
//
// is resolved, but for this to be checked, all codebase (including
// `@oclif/core`) needs to be migrated to ES2022 or higher.
//
// Until then `CONTEXT_FILE_PATH` name is mimicking a `const` while right now it
// is a `let` reassigned inside of `getContextFilePath()`.
export let CONTEXT_FILE_PATH =
  path.resolve(CONTEXT_FILE_LOCATION, CONTEXT_FILENAME) ||
  DEFAULT_CONTEXT_FILE_PATH;

// Sonar recognizes next line as a bug `Promises must be awaited, end with a
// call to .catch, or end with a call to .then with a rejection handler.` but
// due to absence of top-level await in ES6, this bug cannot be fixed without
// migrating the codebase to ES2022 or higher, thus suppressing Sonar analysis
// for this line.
getContextFilePath(); // NOSONAR

export interface IContextFile {
  current?: string;
  readonly store: {
    [name: string]: string;
  };
}

export interface ICurrentContext {
  readonly current: string;
  readonly context: string;
}

export async function initContext(contextFilePath: string) {
  const fileContent: IContextFile = {
    store: {},
  };
  let contextWritePath = '';

  // prettier-ignore
  switch (contextFilePath) {
  /* eslint-disable indent */
    case '.':
      contextWritePath = process.cwd() + path.sep + CONTEXT_FILENAME;
      break;
    case './':
      contextWritePath = REPO_ROOT_PATH + path.sep + CONTEXT_FILENAME;
      break;
    // There are two variants of `~` case because tilde expansion in UNIX
    // systems is not a guaranteed feature - sometimes `~` can return just `~`
    // instead of home directory path.
    // https://stackoverflow.com/questions/491877/how-to-find-a-users-home-directory-on-linux-or-unix#comment17161699_492669
    case os.homedir():
      contextWritePath = os.homedir() + path.sep + CONTEXT_FILENAME;
      break;
    case '~':
      contextWritePath = os.homedir() + path.sep + CONTEXT_FILENAME;
      break;
    default:
      contextWritePath = process.cwd() + path.sep + CONTEXT_FILENAME;
  }

  try {
    await writeFile(contextWritePath, JSON.stringify(fileContent), {
      encoding: 'utf8',
    });
  } catch (e) {
    throw new ContextFileWriteError(contextWritePath);
  }

  return contextWritePath;
}

export async function loadContext(contextName?: string): Promise<string> {
  const fileContent: IContextFile = await loadContextFile();
  if (contextName) {
    const context = fileContent.store[String(contextName)];
    if (!context) {
      throw new ContextNotFoundError(contextName);
    }
    return context;
  } else if (fileContent.current) {
    const context = fileContent.store[fileContent.current];
    if (!context) {
      throw new ContextNotFoundError(fileContent.current);
    }
    return context;
  }
  throw new MissingCurrentContextError();
}

export async function addContext(contextName: string, pathToFile: string) {
  const fileContent: IContextFile = await loadContextFile();

  // If context file already has context name similar to the one specified as
  // an argument, notify user about it (throw `ContextAlreadyExistsError`
  // error) and exit.
  if (fileContent.store.hasOwnProperty.call(fileContent.store, contextName)) {
    throw new ContextAlreadyExistsError(contextName, CONTEXT_FILE_PATH);
  }

  fileContent.store[String(contextName)] = String(pathToFile);
  await saveContextFile(fileContent);
}

export async function removeContext(contextName: string) {
  const fileContent: IContextFile = await loadContextFile();

  if (await isContextFileEmpty(fileContent)) {
    throw new ContextFileEmptyError(CONTEXT_FILE_PATH);
  }
  if (!fileContent.store[String(contextName)]) {
    throw new ContextNotFoundError(contextName);
  }
  if (fileContent.current === contextName) {
    delete fileContent.current;
  }

  delete fileContent.store[String(contextName)];
  await saveContextFile(fileContent);
}

export async function getCurrentContext(): Promise<ICurrentContext> {
  const fileContent: IContextFile = await loadContextFile();

  if (await isContextFileEmpty(fileContent)) {
    throw new ContextFileEmptyError(CONTEXT_FILE_PATH);
  }

  const context = await loadContext();

  return {
    current: fileContent.current as string,
    context,
  };
}

export async function setCurrentContext(contextName: string) {
  const fileContent: IContextFile = await loadContextFile();

  if (await isContextFileEmpty(fileContent)) {
    throw new ContextFileEmptyError(CONTEXT_FILE_PATH);
  }

  if (!fileContent.store[String(contextName)]) {
    throw new ContextNotFoundError(contextName);
  }

  fileContent.current = String(contextName);
  await saveContextFile(fileContent);
}

export async function editContext(contextName: string, pathToFile: string) {
  // The expression is not wrapped in a `try...catch` block and is allowed to
  // throw automatically because it is assumed that `loadContextFile()` works
  // with a 100%-existing valid file in this case, thus if it threw anyway -
  // some REAL error happened and user should know about it.
  const fileContent: IContextFile = await loadContextFile();

  if (await isContextFileEmpty(fileContent)) {
    throw new ContextFileEmptyError(CONTEXT_FILE_PATH);
  }

  fileContent.store[String(contextName)] = String(pathToFile);
  await saveContextFile(fileContent);
}

export async function loadContextFile(): Promise<IContextFile> {
  let fileContent: IContextFile;

  // If the context file cannot be read then it's a 'MissingContextFileError'
  // error.
  try {
    await readFile(CONTEXT_FILE_PATH, { encoding: 'utf8' });
  } catch (e) {
    throw new MissingContextFileError();
  }

  // If the context file cannot be parsed then it's a
  // 'ContextFileWrongFormatError' error.
  try {
    fileContent = JSON.parse(
      await readFile(CONTEXT_FILE_PATH, { encoding: 'utf8' })
    );
  } catch (e) {
    // https://stackoverflow.com/questions/29797946/handling-bad-json-parse-in-node-safely
    throw new ContextFileWrongFormatError(CONTEXT_FILE_PATH);
  }

  // If the context file cannot be validated then it's a
  // 'ContextFileWrongFormatError' error.
  if (!(await isContextFileValid(fileContent))) {
    throw new ContextFileWrongFormatError(CONTEXT_FILE_PATH);
  }

  return fileContent;
}

async function saveContextFile(fileContent: IContextFile) {
  try {
    await writeFile(
      CONTEXT_FILE_PATH,
      JSON.stringify({
        current: fileContent.current,
        store: fileContent.store,
      }),
      { encoding: 'utf8' }
    );
  } catch (e) {
    throw new ContextFileWriteError(CONTEXT_FILE_PATH);
  }
}

async function getRepoRootPath(repoRootPath: string): Promise<string | null> {
  // Asynchronous `fs.exists()` is deprecated, asynchronous `fs.stat()`
  // introduces race condition, thus synchronous functions are used.

  let pathToCheck = `${repoRootPath}${path.sep}.git`;

  // If directory where `init` was requested in, happens to contain `.git`
  // directory, then it surely is a root of repository, no need to search
  // further and `REPO_ROOT_PATH` will remain as it was.
  if (existsSync(pathToCheck) && lstatSync(pathToCheck).isDirectory()) {
    return null;
  }

  // Directory where `init` was requested in, did not happen to contain `.git`
  // directory, so preparation for iterating through array of filesystem paths
  // is started.
  const repoRootPathArray = repoRootPath.split(path.sep);

  // Last element in array is thrown away because it is already known that it
  // does not contain directory `.git`.
  repoRootPathArray.pop();

  // Backwards search of the array of filesystem paths will now be performed.
  let i = repoRootPathArray.length - 1;

  while (i > 0) {
    pathToCheck = `${repoRootPathArray.join(path.sep)}${path.sep}.git`;

    if (existsSync(pathToCheck) && lstatSync(pathToCheck).isDirectory()) {
      REPO_ROOT_PATH = repoRootPathArray.join(path.sep);
      return REPO_ROOT_PATH;
    }

    // Last (`0th`) element is an empty string, so if directory `.git` was not
    // found on 1st element (last actual directory in filesystem), the search
    // does not need to continue and `REPO_ROOT_PATH` will remain having the
    // value of current (where `init` was requested in) directory.
    if (i === 1) {
      return null;
    }

    repoRootPathArray.pop();

    i--;
  }
  return null;
}

async function getContextFilePath(): Promise<string | null> {
  const currentPath = process
    .cwd()
    .slice(REPO_ROOT_PATH.length + 1)
    .split(path.sep);
  currentPath.unshift(REPO_ROOT_PATH);

  for (let i = currentPath.length; i >= 0; i--) {
    const currentPathString = currentPath[0]
      ? currentPath.join(path.sep) + path.sep + CONTEXT_FILENAME
      : os.homedir() + path.sep + CONTEXT_FILENAME;

    // This `try...catch` is a part of `for` loop and is used only to swallow
    // errors if the file does not exist or cannot be read, to continue
    // uninterrupted execution of the loop.
    try {
      // If a file is found which can be read and passed validation as a
      // legitimate context file, then it is considered a legitimate context
      // file indeed.
      const fileContent = JSON.parse(
        //we do not use await readFile because getContextFilePath cannot be called inside async function
        readFileSync(currentPathString, {encoding: 'utf8'})
      );
      if (
        fileContent &&
        (await isContextFileValid(fileContent as unknown as IContextFile))
      ) {
        CONTEXT_FILE_PATH = currentPathString;
        return CONTEXT_FILE_PATH;
      }
    } catch (e) {} // eslint-disable-line

    currentPath.pop();
  }
  return null;
}

async function isContextFileValid(fileContent: IContextFile): Promise<boolean> {
  // Validation of context file's format against interface `IContextFile`.
  return (
    [1, 2].includes(Object.keys(fileContent).length) &&
    fileContent.hasOwnProperty.call(fileContent, 'store') &&
    !Array.from(Object.keys(fileContent.store)).find(
      (elem) => typeof elem !== 'string'
    ) &&
    !Array.from(Object.values(fileContent.store)).find(
      (elem) => typeof elem !== 'string'
    )
  );
}

export async function isContextFileEmpty(
  fileContent: IContextFile
): Promise<boolean> {
  // If context file contains only one empty property `store` then the whole
  // context file is considered empty.
  return (
    fileContent &&
    Object.keys(fileContent).length === 1 &&
    Object.keys(fileContent.store).length === 0
  );
}
