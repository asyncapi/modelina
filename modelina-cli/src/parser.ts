import { AvroSchemaParser } from '@asyncapi/avro-schema-parser';
import { OpenAPISchemaParser } from '@asyncapi/openapi-schema-parser';
import { Parser, convertToOldAPI } from '@asyncapi/parser/cjs';
import { RamlDTSchemaParser } from '@asyncapi/raml-dt-schema-parser';
import { Flags } from '@oclif/core';
import { ProtoBuffSchemaParser } from '@asyncapi/protobuf-schema-parser';
import { getDiagnosticSeverity } from '@stoplight/spectral-core';
import { OutputFormat } from '@stoplight/spectral-cli/dist/services/config';
import { html, json, junit, pretty, stylish, teamcity, text } from '@stoplight/spectral-formatters';

import type { Diagnostic } from '@asyncapi/parser/cjs';
import type Command from './base';
import type { Specification } from './models/SpecificationFile';

export type SeveritytKind = 'error' | 'warn' | 'info' | 'hint';

export { convertToOldAPI };

const parser = new Parser({
  __unstable: {
    resolver: {
      cache: false,
    }
  }
});

parser.registerSchemaParser(AvroSchemaParser());
parser.registerSchemaParser(OpenAPISchemaParser());
parser.registerSchemaParser(RamlDTSchemaParser());
parser.registerSchemaParser(ProtoBuffSchemaParser());

export interface ValidationFlagsOptions {
  logDiagnostics?: boolean;
}

export function validationFlags({ logDiagnostics = true }: ValidationFlagsOptions = {}) {
  return {
    'log-diagnostics': Flags.boolean({
      description: 'log validation diagnostics or not',
      default: logDiagnostics,
      allowNo: true,
    }),
    'diagnostics-format': Flags.enum({
      description: 'format to use for validation diagnostics',
      options: Object.values(OutputFormat),
      default: OutputFormat.STYLISH,
    }),
    'fail-severity': Flags.enum<SeveritytKind>({
      description: 'diagnostics of this level or above will trigger a failure exit code',
      options: ['error', 'warn', 'info', 'hint'],
      default: 'error',
    }),
  };
}

interface ValidateOptions {
  'log-diagnostics'?: boolean;
  'diagnostics-format'?: `${OutputFormat}`;
  'fail-severity'?: SeveritytKind;
}

export async function validate(command: Command, specFile: Specification, options: ValidateOptions = {}) {
  const diagnostics = await parser.validate(specFile.text(), { source: specFile.getSource() });
  return logDiagnostics(diagnostics, command, specFile, options);
}

export async function parse(command: Command, specFile: Specification, options: ValidateOptions = {}) {
  const { document, diagnostics } = await parser.parse(specFile.text(), { source: specFile.getSource() });
  const status = logDiagnostics(diagnostics, command, specFile, options);
  return { document, diagnostics, status };
}

function logDiagnostics(diagnostics: Diagnostic[], command: Command, specFile: Specification, options: ValidateOptions = {}): 'valid' | 'invalid' {
  const logDiagnostics = options['log-diagnostics'];
  const failSeverity = options['fail-severity'] ?? 'error';
  const diagnosticsFormat = options['diagnostics-format'] ?? 'stylish';

  const sourceString = specFile.toSourceString();
  if (diagnostics.length) {
    if (hasFailSeverity(diagnostics, failSeverity)) {
      if (logDiagnostics) {
        command.logToStderr(`\n${sourceString} and/or referenced documents have governance issues.`);
        command.logToStderr(formatOutput(diagnostics, diagnosticsFormat, failSeverity));
      }
      return 'invalid';
    }

    if (logDiagnostics) {
      command.log(`\n${sourceString} is valid but has (itself and/or referenced documents) governance issues.`);
      command.log(formatOutput(diagnostics, diagnosticsFormat, failSeverity));
    }
  } else if (logDiagnostics) {
    command.log(`\n${sourceString} is valid! ${sourceString} and referenced documents don't have governance issues.`);
  }

  return 'valid';
}

export function formatOutput(diagnostics: Diagnostic[], format: `${OutputFormat}`, failSeverity: SeveritytKind) {
  const options = { failSeverity: getDiagnosticSeverity(failSeverity) };
  switch (format) {
  case 'stylish': return stylish(diagnostics, options);
  case 'json': return json(diagnostics, options);
  case 'junit': return junit(diagnostics, options);
  case 'html': return html(diagnostics, options);
  case 'text': return text(diagnostics, options);
  case 'teamcity': return teamcity(diagnostics, options);
  case 'pretty': return pretty(diagnostics, options);
  default: return stylish(diagnostics, options);
  }
}

function hasFailSeverity(diagnostics: Diagnostic[], failSeverity: SeveritytKind) {
  const diagnosticSeverity = getDiagnosticSeverity(failSeverity);
  return diagnostics.some(diagnostic => diagnostic.severity <= diagnosticSeverity);
}
