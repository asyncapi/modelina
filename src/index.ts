export * from './generators';
export * from './helpers';
export * from './models';
export * from './processors';
export type { ModelLoggingInterface } from './utils';
export { Logger } from './utils';
