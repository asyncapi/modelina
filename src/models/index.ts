export * from './CommonModel';
export * from './RenderOutput';
export * from './OutputModel';
export * from './Preset';
export * from './ProcessorOptions';
export * from './AsyncapiV2Schema';
export * from './Draft7Schema';
export * from './Draft6Schema';
export * from './Draft4Schema';
export * from './SwaggerV2Schema';
export * from './OpenapiV3Schema';
export * from './MetaModel';
export * from './ConstrainedMetaModel';
export * from './InputMetaModel';
