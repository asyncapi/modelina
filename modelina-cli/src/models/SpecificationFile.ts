import { promises as fs } from 'fs';
import path from 'path';
import { URL } from 'url';
import fetch from 'node-fetch';
import yaml from 'js-yaml';
import { loadContext } from './Context';
import { ErrorLoadingSpec } from '../errors/specification-file';
import { MissingContextFileError } from '../errors/context-error';

const { readFile, lstat } = fs;
const allowedFileNames: string[] = [
  'asyncapi.json',
  'asyncapi.yml',
  'asyncapi.yaml'
];
const TYPE_CONTEXT_NAME = 'context-name';
const TYPE_FILE_PATH = 'file-path';
const TYPE_URL = 'url-path';

export class Specification {
  private readonly spec: string;
  private readonly filePath?: string;
  private readonly fileURL?: string;
  private readonly kind?: 'file' | 'url';

  constructor(spec: string, options: { filepath?: string, fileURL?: string } = {}) {
    this.spec = spec;
    if (options.filepath) {
      this.filePath = options.filepath;
      this.kind = 'file';
    } else if (options.fileURL) {
      this.fileURL = options.fileURL;
      this.kind = 'url';
    }
  }

  isAsyncAPI3() {
    const jsObj = this.toJson();
    return jsObj.asyncapi === '3.0.0';
  }

  toJson(): Record<string, any> {
    try {
      return yaml.load(this.spec, {json: true}) as Record<string, any>;
    } catch (e) {
      return JSON.parse(this.spec);
    }
  }

  text() {
    return this.spec;
  }

  getFilePath() {
    return this.filePath;
  }

  getFileURL() {
    return this.fileURL;
  }

  getKind() {
    return this.kind;
  }

  getSource() {
    return this.getFilePath() ?? this.getFileURL();
  }

  toSourceString() {
    if (this.kind === 'file') {
      return `File ${this.filePath}`;
    }
    return `URL ${this.fileURL}`;
  }

  static async fromFile(filepath: string) {
    let spec;
    try {
      spec = await readFile(filepath, { encoding: 'utf8' });
    } catch (error) {
      throw new ErrorLoadingSpec('file', filepath);
    }
    return new Specification(spec, { filepath });
  }

  static async fromURL(URLpath: string) {
    let response;
    try {
      response = await fetch(URLpath, { method: 'GET' });
      if (!response.ok) {
        throw new ErrorLoadingSpec('url', URLpath);
      }
    } catch (error) {
      throw new ErrorLoadingSpec('url', URLpath);
    }
    return new Specification(await response?.text() as string, { fileURL: URLpath });
  }
}

export default class SpecificationFile {
  private readonly pathToFile: string;

  constructor(filePath: string) {
    this.pathToFile = filePath;
  }

  getPath(): string {
    return this.pathToFile;
  }

  async read(): Promise<string> {
    return readFile(this.pathToFile, { encoding: 'utf8' });
  }
}

interface LoadType {
  file?: boolean
  url?: boolean
  context?: boolean
}

/* eslint-disable sonarjs/cognitive-complexity */
export async function load(filePathOrContextName?: string, loadType?: LoadType): Promise<Specification> { // NOSONAR
  if (filePathOrContextName) {
    if (loadType?.file) { return Specification.fromFile(filePathOrContextName); }
    if (loadType?.context) { return loadFromContext(filePathOrContextName); }
    if (loadType?.url) { return Specification.fromURL(filePathOrContextName); }

    const type = await nameType(filePathOrContextName);
    if (type === TYPE_CONTEXT_NAME) {
      return loadFromContext(filePathOrContextName);
    }

    if (type === TYPE_URL) {
      return Specification.fromURL(filePathOrContextName);
    }
    await fileExists(filePathOrContextName);
    return Specification.fromFile(filePathOrContextName);
  }

  try {
    return await loadFromContext();
  } catch (e) {
    const autoDetectedSpecFile = await detectSpecFile();
    if (autoDetectedSpecFile) {
      return Specification.fromFile(autoDetectedSpecFile);
    }

    if (e instanceof MissingContextFileError) {
      throw new ErrorLoadingSpec();
    }

    throw e;
  }
}

export async function nameType(name: string): Promise<string> {
  if (name.startsWith('.')) {
    return TYPE_FILE_PATH;
  }

  try {
    if (await fileExists(name)) {
      return TYPE_FILE_PATH;
    }
    return TYPE_CONTEXT_NAME;
  } catch (e) {
    if (await isURL(name)) { return TYPE_URL; }
    return TYPE_CONTEXT_NAME;
  }
}

export async function isURL(urlpath: string): Promise<boolean> {
  try {
    const url = new URL(urlpath);
    return url.protocol === 'http:' || url.protocol === 'https:';
  } catch (error) {
    return false;
  }
}

export async function fileExists(name: string): Promise<boolean> {
  try {
    if ((await lstat(name)).isFile()) {
      return true;
    }
    
    const extension = name.split('.')[1];
    
    const allowedExtenstion=['yml','yaml','json'];

    if (!allowedExtenstion.includes(extension)) {
      throw new ErrorLoadingSpec('invalid file',name);
    }
   
    throw new ErrorLoadingSpec('file', name);
  } catch (e) {
    throw new ErrorLoadingSpec('file', name);
  }
}

async function loadFromContext(contextName?: string): Promise<Specification> {
  try {
    const context = await loadContext(contextName);
    return Specification.fromFile(context);
  } catch (error) {
    if (error instanceof MissingContextFileError) {throw new ErrorLoadingSpec();}
    throw error;
  }
}

async function detectSpecFile(): Promise<string | undefined> {
  const existingFileNames = await Promise.all(allowedFileNames.map(async filename => {
    try {
      const exists = await fileExists(path.resolve(process.cwd(), filename));
      return exists ? filename : undefined;
    } catch (e) {
      // We did our best...
    }
  }));
  return existingFileNames.find(filename => filename !== undefined);
}

