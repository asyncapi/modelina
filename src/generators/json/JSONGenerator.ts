import { 
  AbstractGenerator, 
  CommonGeneratorOptions,
  defaultGeneratorOptions
} from '../AbstractGenerator';
import { CommonModel, CommonInputModel, RenderOutput } from '../../models';
import { TypeHelpers, ModelKind, CommonNamingConvention,CommonNamingConventionImplementation } from '../../helpers';
import { Logger } from '../../utils';
import { JSONRenderer } from './JSONRenderer';

export interface JSONOptions extends CommonGeneratorOptions {
  schemaType? : 'draft-04' | 'draft-06' | 'draft-07';
  namingConvention?: CommonNamingConvention
}

export class JSONGenerator extends AbstractGenerator<JSONOptions, any> {
  static defaultOptions:JSONOptions = {
    ...defaultGeneratorOptions,
    namingConvention: CommonNamingConventionImplementation,
  }

  constructor(
    options: JSONOptions = JSONGenerator.defaultOptions,
  ) {
    super('JSON', JSONGenerator.defaultOptions, options);
  }

  /**
   * Render a scattered model, where the source code and library and model dependencies are separated.
   * 
   * @param model 
   * @param inputModel 
   */
  public render(model: CommonModel, inputModel: CommonInputModel): Promise<RenderOutput> {
    const kind = TypeHelpers.extractKind(model);

    if (kind === ModelKind.ENUM || kind === ModelKind.ARRAY || kind === ModelKind.UNION) {
      return this.renderArray(model, inputModel);
    }

    if (kind === ModelKind.OBJECT) {
      return this.renderObject(model, inputModel);
    }

    Logger.warn(`JS generator, cannot generate model for '${model.$id} since type: ${kind}'`);
    return Promise.resolve(RenderOutput.toRenderOutput({result: '', renderedName: '', dependencies: []}));
  }

  /**
   * Render a complete model result where the model code, library and model dependencies are all bundled appropriately.
   * 
   * For Java you need to specify which package the model is placed under.
   * 
   * @param model 
   * @param inputModel 
   * @param options used to render the full output
   */
  renderCompleteModel(model: CommonModel, inputModel: CommonInputModel, options: JSONOptions): Promise<RenderOutput> {
    console.log(options);
    return this.render(model,inputModel);
  }

  async renderObject(model: CommonModel, inputModel: CommonInputModel): Promise<RenderOutput> {
    const renderer = new JSONRenderer(this.options, this, [], model, inputModel);
    const renderedType = await renderer.renderType(model);
    const renderedName = renderer.nameType(model.$id, model);
    const renderProperties = renderer.renderAllProperties(model);
    const result = JSON.stringify({
      $id: renderedName,
      type: renderedType,
      ...renderProperties
    }, null, 4);
    return RenderOutput.toRenderOutput({result, renderedName, dependencies: renderer.dependencies});
  }

  async renderArray(model: CommonModel, inputModel: CommonInputModel): Promise<RenderOutput> {
    const renderer = new JSONRenderer(this.options, this, [], model, inputModel);
    const renderedContents = await renderer.renderArrayItems(model);
    const renderedName = renderer.nameType(model.$id, model);
    const result = JSON.stringify({
      $id: renderedName,
      type: 'array',
      ...renderedContents 
    }, null , 4);
    return RenderOutput.toRenderOutput({result, renderedName, dependencies: renderer.dependencies});
  }
}
