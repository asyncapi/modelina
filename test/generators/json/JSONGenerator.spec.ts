import { JSONGenerator } from '../../../src/generators/json/JSONGenerator';
import { CommonInputModel, CommonModel } from '../../../src/models';

describe('JSONGenerator', () => {
  let generator: JSONGenerator;
  beforeEach(() => {
    generator = new JSONGenerator();
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });

  test('Should render JSON schema for type object', async () => {
    const doc = {
      $id: 'Address',
      type: 'object',
      properties: {
        enum: { type: 'string' },
        reservedEnum: { type: 'string' }
      },
      additionalProperties: {
        type: 'string'
      },
      patternProperties: {
        '^S_': { type: 'string' },
        '^I_': { type: 'integer' }
      }
    };

    const inputModel = await generator.process(doc);
    const model = inputModel.models['Address'];
    const schema = await generator.render(model, inputModel);
    expect(schema.result).toMatchSnapshot();
    expect(schema.dependencies).toEqual([]);
  });

  test('Should render JSON schema for array object', async () => {
    const doc = {
      $id: 'TypeArray',
      type: 'array',
      items: {
        $id: 'StringArray',
        type: ['string', 'number', 'boolean'],
      }
    };

    const inputModel = await generator.process(doc);
    const model = inputModel.models['TypeArray'];
    const schema = await generator.render(model, inputModel);
    expect(schema.result).toMatchSnapshot();
    expect(schema.dependencies).toEqual([]);
  });
});

// const doc = {
//   $id: 'Address',
//   type: 'object',
//   properties: {
//     enum: { type: 'string' },
//     reservedEnum: { type: 'string' }
//   },
//   additionalProperties: {
//     type: 'string'
//   },
//   patternProperties: {
//     '^S_': { type: 'string' },
//     '^I_': { type: 'integer' }
//   }
// };

// const doc = {
//   $id: 'Address',
//   type: 'object',
//   properties: {
//     street_name: { type: 'string' },
//     city: { type: 'string', description: 'City description' },
//     state: { type: 'string' },
//     house_number: { type: 'number' },
//     marriage: { type: 'boolean', description: 'Status if marriage live in given house' },
//     members: { oneOf: [{ type: 'string' }, { type: 'number' }, { type: 'boolean' }], },
//     array_type: { type: 'array', items: [{ type: 'string' }, { type: 'number' }] },
//     other_model: { type: 'object', $id: 'OtherModel', properties: { street_name: { type: 'string' } } },
//   },
//   patternProperties: {
//     '^S(.?*)test&': {
//       type: 'string'
//     }
//   },
//   required: ['street_name', 'city', 'state', 'house_number', 'array_type'],
// };

// const doc = {
//   $id: 'TypeArray',
//   type: 'array',
//   items: {
//     $id: 'StringArray',
//     type: ['string', 'number', 'boolean'],
//   }
// };

// const generator = new JSONGenerator();
// generator.generate(doc).then((output) => {
//   console.log(output);
// })
