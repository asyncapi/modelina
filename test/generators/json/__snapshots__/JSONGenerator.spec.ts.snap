// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONGenerator Should render JSON schema for array object 1`] = `
"{
    \\"$id\\": \\"TypeArray\\",
    \\"type\\": \\"array\\",
    \\"additionalItems\\": [
        \\"object\\",
        \\"string\\",
        \\"number\\",
        \\"array\\",
        \\"boolean\\",
        \\"null\\",
        \\"integer\\"
    ],
    \\"items\\": {
        \\"type\\": [
            \\"string\\",
            \\"number\\",
            \\"boolean\\"
        ],
        \\"$id\\": \\"StringArray\\"
    }
}"
`;

exports[`JSONGenerator Should render JSON schema for type object 1`] = `
"{
    \\"$id\\": \\"Address\\",
    \\"type\\": \\"object\\",
    \\"properties\\": {
        \\"enum\\": {
            \\"type\\": \\"string\\"
        },
        \\"reservedEnum\\": {
            \\"type\\": \\"string\\"
        }
    },
    \\"additionalProperties\\": {
        \\"type\\": \\"string\\"
    },
    \\"patternProperties\\": {
        \\"^S_\\": {
            \\"type\\": \\"string\\"
        },
        \\"^I_\\": {
            \\"type\\": \\"integer\\"
        }
    }
}"
`;
